File: apps/desktop-pdv/src/pdv/sync/pull_mapping.py
"""Como uma linha de cadastro da nuvem vira uma linha no banco do caixa.

Até aqui o pull fazia `INSERT` com **todas** as chaves que a nuvem mandasse,
com os nomes das colunas vindos da resposta interpolados no SQL. As duas pontas
nunca tiveram o mesmo schema, e o pull nunca aplicou uma linha sequer:

* `users` chega com `server_seq`, que o caixa não tem → "no column named";
* `products` chega sem `store_id`, obrigatório no caixa (o catálogo da nuvem é
  da rede inteira; o do caixa é da loja);
* receitas e insumos têm modelos diferentes (`yield_grams` × `base_qty_g` +
  `yield_factor`, `unit_cost_cents` × `avg_cost_cents_per_kg`).

Aqui cada tabela diz o que aceita. Coluna nova na nuvem é ignorada em vez de
derrubar o caixa, nome de coluna nunca vem da rede, e o que o caixa precisa e
a nuvem não manda é preenchido pela configuração do terminal.

O que **não** é aplicado, e por quê
-----------------------------------

Receitas, linhas de receita e insumos ficam de fora até os modelos convergirem.
Converter `yield_grams` em `base_qty_g`/`yield_factor`, ou custo por unidade em
custo por quilo, é decidir a regra de baixa de estoque por adivinhação — e a
baixa errada aparece no CMV do mês, não no dia. E o saldo de insumo é do caixa
(ele é que registra os movimentos): aceitar o saldo da nuvem por LWW apagaria
movimentos locais ainda não sincronizados.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any

from pdv.config import AppConfig


def _int(value: Any) -> int:
    # O driver da nuvem serializa BIGINT como texto ("700") para não perder
    # precisão no JSON; aqui volta a ser inteiro.
    return int(str(value))


def _flag(value: Any) -> int:
    if isinstance(value, str):
        return 1 if value.strip().lower() in ("1", "true", "t", "yes") else 0
    return 1 if value else 0


def _text(value: Any) -> str:
    return str(value)


@dataclass(frozen=True, slots=True)
class PullMapping:
    """O contrato de uma tabela de cadastro."""

    #: Colunas copiadas da nuvem quando presentes, com a conversão de cada uma.
    columns: dict[str, Callable[[Any], Any]]
    #: Sem estas, a linha é descartada — melhor que gravar cadastro pela metade.
    required: tuple[str, ...]
    #: O que o caixa exige e a nuvem não manda.
    fill: Callable[[AppConfig], dict[str, Any]] = field(default=lambda _c: {})


PULL_MAPPINGS: dict[str, PullMapping] = {
    "users": PullMapping(
        columns={
            "id": _text,
            "tenant_id": _text,
            "name": _text,
            "login": _text,
            "role": _text,
            "pin_hash": _text,
            "max_discount_percent": _text,
            "can_authorize": _flag,
            "is_active": _flag,
            "updated_at": _text,
        },
        required=("id", "tenant_id", "name", "login", "role", "updated_at"),
    ),
    "products": PullMapping(
        columns={
            "id": _text,
            "tenant_id": _text,
            "sku": _text,
            "barcode": _text,
            "name": _text,
            "category": _text,
            "pricing_mode": _text,
            "price_cents": _int,
            "tare_grams": _int,
            "is_active": _flag,
            "updated_at": _text,
            "server_seq": _int,
        },
        required=("id", "tenant_id", "sku", "name", "pricing_mode", "price_cents", "updated_at"),
        # O catálogo da nuvem vale para a rede; no caixa, o produto é desta loja.
        fill=lambda config: {"store_id": config.store_id},
    ),
}

#: Tabelas que a nuvem oferece e o caixa ainda não sabe aplicar com segurança.
NOT_APPLIED: dict[str, str] = {
    "recipes": "modelo de rendimento diferente (yield_grams × base_qty_g/yield_factor)",
    "recipe_lines": "quantidade por base diferente (quantity_mg × qty_per_base_mg)",
    "inventory_items": "o saldo é do caixa; custo em unidades diferentes",
}


def map_row(
    table: str, row: dict[str, Any], config: AppConfig
) -> dict[str, Any] | None:
    """A linha pronta para o banco do caixa, ou `None` se ela não serve.

    Linha de OUTRO tenant é descartada aqui também. A nuvem já filtra; conferir
    de novo custa uma comparação, e é o que impede um cadastro de outra rede —
    por defeito ou por ataque — de virar login válido neste caixa.
    """
    mapping = PULL_MAPPINGS[table]
    if str(row.get("tenant_id") or "") != config.tenant_id:
        return None
    if any(row.get(name) in (None, "") for name in mapping.required):
        return None

    values: dict[str, Any] = {}
    for name, convert in mapping.columns.items():
        if name not in row:
            continue
        raw = row[name]
        if raw is None:
            # Nulo vindo da nuvem é "sem valor", não "apague": fica o padrão do
            # caixa na inserção e o valor atual na atualização. Gravar NULL
            # violaria `NOT NULL DEFAULT` do lado de cá e derrubaria a tabela.
            continue
        try:
            values[name] = convert(raw)
        except (TypeError, ValueError):
            return None
    values.update(mapping.fill(config))
    return values


__all__ = ["NOT_APPLIED", "PULL_MAPPINGS", "PullMapping", "map_row"]

File: apps/pdv-net/src/Pdv.Data/Sync/PullMapping.cs
using System.Globalization;
using System.Text.Json;
using System.Text.RegularExpressions;
using Pdv.Core.Audit;

namespace Pdv.Data.Sync;

/// <summary>
/// Como uma linha de cadastro da nuvem vira uma linha no banco do caixa — o
/// <c>pull_mapping</c> do Python, conferido caso a caso por <c>contracts/sync.json</c>.
/// </summary>
/// <remarks>
/// <para>
/// Cada tabela diz o que aceita. Coluna nova na nuvem é ignorada em vez de
/// derrubar o caixa, nome de coluna nunca vem da rede, e o que o caixa exige e
/// a nuvem não manda (a loja do produto) vem da identidade do terminal.
/// </para>
/// <para>
/// Receitas, linhas de receita e insumos ficam de fora até os modelos
/// convergirem: converter rendimento e custo por adivinhação erra a baixa de
/// estoque, e o erro aparece no CMV do mês.
/// </para>
/// </remarks>
public static partial class PullMapping
{
    /// <summary>Cadastros que a nuvem oferece, na ordem em que o caixa pede.</summary>
    public static readonly IReadOnlyList<string> PullableTables = ["products", "recipes", "recipe_lines", "inventory_items", "users"];

    /// <summary>O que a nuvem oferece e o caixa ainda não aplica com segurança — e por quê.</summary>
    public static readonly IReadOnlyDictionary<string, string> NotApplied = new Dictionary<string, string>
    {
        ["recipes"] = "modelo de rendimento diferente (yield_grams × base_qty_g/yield_factor)",
        ["recipe_lines"] = "quantidade por base diferente (quantity_mg × qty_per_base_mg)",
        ["inventory_items"] = "o saldo é do caixa; custo em unidades diferentes",
    };

    private sealed record Mapping(
        IReadOnlyDictionary<string, Func<JsonElement, object?>> Columns,
        IReadOnlyList<string> Required,
        Func<string, IReadOnlyDictionary<string, object>> Fill);

    private static readonly IReadOnlyDictionary<string, Mapping> Mappings = new Dictionary<string, Mapping>
    {
        ["users"] = new(
            new Dictionary<string, Func<JsonElement, object?>>
            {
                ["id"] = Text, ["tenant_id"] = Text, ["name"] = Text, ["login"] = Text, ["role"] = Text,
                ["pin_hash"] = Text, ["max_discount_percent"] = Text, ["can_authorize"] = Flag,
                ["is_active"] = Flag, ["updated_at"] = Text,
            },
            ["id", "tenant_id", "name", "login", "role", "updated_at"],
            _ => new Dictionary<string, object>()),
        ["products"] = new(
            new Dictionary<string, Func<JsonElement, object?>>
            {
                ["id"] = Text, ["tenant_id"] = Text, ["sku"] = Text, ["barcode"] = Text, ["name"] = Text,
                ["category"] = Text, ["pricing_mode"] = Text, ["price_cents"] = Int, ["tare_grams"] = Int,
                ["is_active"] = Flag, ["updated_at"] = Text, ["server_seq"] = Int,
            },
            ["id", "tenant_id", "sku", "name", "pricing_mode", "price_cents", "updated_at"],
            // O catálogo da nuvem vale para a rede; no caixa, o produto é desta loja.
            storeId => new Dictionary<string, object> { ["store_id"] = storeId }),
    };

    public static IEnumerable<string> MappedTables => Mappings.Keys;

    public static bool IsMapped(string table) => Mappings.ContainsKey(table);

    /// <summary>
    /// A linha pronta para o banco do caixa, ou <c>null</c> se ela não serve:
    /// outro tenant, campo obrigatório faltando ou valor que não converte.
    /// </summary>
    /// <remarks>
    /// A nuvem já filtra por tenant; conferir de novo custa uma comparação e é
    /// o que impede um cadastro de outra rede de virar login neste caixa. Nulo
    /// vindo da nuvem é "sem valor", não "apague": fica o padrão do caixa.
    /// </remarks>
    public static IReadOnlyDictionary<string, object>? MapRow(string table, JsonElement row, string tenantId, string storeId)
    {
        var mapping = Mappings[table];
        if (row.ValueKind != JsonValueKind.Object) return null;
        if (!row.TryGetProperty("tenant_id", out var tenant) || PyStr(tenant) != tenantId) return null;
        foreach (var name in mapping.Required)
        {
            if (!row.TryGetProperty(name, out var value) || value.ValueKind == JsonValueKind.Null ||
                (value.ValueKind == JsonValueKind.String && value.GetString()!.Length == 0))
            {
                return null;
            }
        }

        var values = new Dictionary<string, object>(StringComparer.Ordinal);
        foreach (var (name, convert) in mapping.Columns)
        {
            if (!row.TryGetProperty(name, out var raw) || raw.ValueKind == JsonValueKind.Null) continue;
            var converted = convert(raw);
            if (converted is null) return null;
            values[name] = converted;
        }
        foreach (var (name, value) in mapping.Fill(storeId)) values[name] = value;
        return values;
    }

    // -- as conversões do Python -------------------------------------------------

    /// <summary><c>str(value)</c> do Python sobre o que o <c>json.loads</c> produz.</summary>
    private static object? Text(JsonElement value) => PyStr(value);

    private static string? PyStr(JsonElement value) => value.ValueKind switch
    {
        JsonValueKind.String => value.GetString(),
        JsonValueKind.True => "True",
        JsonValueKind.False => "False",
        JsonValueKind.Number => IsIntegerLiteral(value.GetRawText())
            ? System.Numerics.BigInteger.Parse(value.GetRawText(), CultureInfo.InvariantCulture).ToString(CultureInfo.InvariantCulture)
            : PythonFloat.Repr(value.GetDouble()),
        JsonValueKind.Null => "None",
        _ => value.GetRawText(),
    };

    /// <summary>
    /// <c>int(str(value))</c>: o driver da nuvem manda BIGINT como texto
    /// (<c>"700"</c>). Aceita espaço em volta, sinal e <c>_</c> entre dígitos,
    /// como o Python; recusa <c>"12.0"</c>, número com fração e booleano.
    /// </summary>
    private static object? Int(JsonElement value)
    {
        var text = value.ValueKind switch
        {
            JsonValueKind.String => value.GetString()!,
            JsonValueKind.Number when IsIntegerLiteral(value.GetRawText()) => value.GetRawText(),
            _ => null,
        };
        if (text is null) return null;
        var match = PythonIntLiteral().Match(text);
        if (!match.Success) return null;
        return long.TryParse(match.Groups["number"].Value.Replace("_", ""), NumberStyles.AllowLeadingSign,
            CultureInfo.InvariantCulture, out var parsed)
            ? parsed
            : null;
    }

    /// <summary>
    /// Texto: <c>1/true/t/yes</c>, sem caixa e sem espaço em volta. Resto: a
    /// "verdade" do Python — zero, vazio e falso são 0.
    /// </summary>
    private static object? Flag(JsonElement value) => value.ValueKind switch
    {
        JsonValueKind.String => value.GetString()!.Trim().ToLowerInvariant() is "1" or "true" or "t" or "yes" ? 1 : 0,
        JsonValueKind.True => 1,
        JsonValueKind.False => 0,
        JsonValueKind.Number => value.GetDouble() != 0 ? 1 : 0,
        JsonValueKind.Array => value.GetArrayLength() > 0 ? 1 : 0,
        JsonValueKind.Object => value.EnumerateObject().Any() ? 1 : 0,
        _ => 0,
    };

    private static bool IsIntegerLiteral(string raw) => raw.IndexOfAny(['.', 'e', 'E']) < 0;

    [GeneratedRegex(@"^\s*(?<number>[+-]?[0-9]+(?:_[0-9]+)*)\s*$")]
    private static partial Regex PythonIntLiteral();
}
